.github/workflows/publish.yml

```diff
@@ -14,38 +17,95 @@ on:
+    # Windows. So, we upload to the Github Actions artifacts, then download it on Linux in a later
+    # step to upload to the module registry.
+    - name: upload-module
+      uses: actions/upload-artifact@v4
+      with:
+        name: windows-2019-module-${{ github.sha }}
+        path: ./module.tar.gz
+
+  build-unix:
+    name: Build for Unix
+    strategy:
+      matrix:
+        build_target:
+          - runs_on: ubuntu-latest
+          - runs_on: ubuntu-24.04-arm
+          - runs_on: macos-latest
+          - runs_on: macos-13
+    runs-on: ${{ matrix.build_target.runs_on }}
     steps:
       - name: Checkout
         uses: actions/checkout@v4
 
-      - uses: actions/cache@v3
+      - name: Restore cached libraries
+        uses: actions/cache@v3
         with:
           path: ~/.conan2
-          key: ${{runner.os}}-conan-${{ hashFiles('**/conanfile.py') }}
+          key: ${{ runner.os }}-${{ matrix.build_target.runs_on }}-conan-${{ hashFiles('**/conanfile.py') }}
           restore-keys: |
-            ${{ runner.os }}-conan-
+            ${{ runner.os }}-${{ matrix.build_target.runs_on }}-conan-
+      - name: Setup and build
+        run: |
+          make setup
+          make module.tar.gz
+      - name: Store module temporarily
+        uses: actions/upload-artifact@v4
+        with:
+          name: ${{ matrix.build_target.runs_on }}-module-${{ github.sha }}
+          path: ./module.tar.gz
 
-      - name: Build and Publish
-        uses: viamrobotics/build-action@v1
+  publish:
+    name: Build and Upload module
+    needs: [build-windows, build-unix]
+    strategy:
+      matrix:
+        build_target:
+          - platform: linux/amd64
+            runs_on: ubuntu-latest
+          - platform: linux/arm64
+            runs_on: ubuntu-24.04-arm
+          - platform: darwin/arm64
+            runs_on: macos-latest
+          - platform: darwin/amd64
+            runs_on: macos-13
+          - platform: windows/amd64
+            runs_on: windows-2019
+    runs-on: ubuntu-latest
+    steps:
+      - name: Install Viam CLI
+        run: |
+          sudo curl -o /usr/local/bin/viam https://storage.googleapis.com/packages.viam.com/apps/viam-cli/viam-cli-stable-linux-amd64
+          sudo chmod a+rx /usr/local/bin/viam
+      - name: Download build artifact
+        uses: actions/download-artifact@v4
         with:
-          version: ${{ github.ref_name }}
-          ref: ${{ github.sha }}
-          key-id: ${{ secrets.viam_key_id }}
-          key-value: ${{ secrets.viam_key_value }}
+          name: ${{ matrix.build_target.runs_on }}-module-${{ github.sha }}
+          path: .
+      - name: Publish build to Viam registry
+        run: |
+          #viam login api-key --key-id ${{ secrets.viam_key_id }} --key ${{ secrets.viam_key_value }}
+          #viam module upload --platform ${{ matrix.build_target.platform }} --version ${{ github.ref_name }} --upload module.tar.gz
+          ls -la .
```

Comment:
Reminder: delete this line and uncomment the previous two before merging. We don't have an API key when running on non-merged commits, so I can't test out logging into Viam and uploading the module. I sure hope that part works...